Cluster API Security Self Assessment

PushkarJ opened the pull request:
## Description

SIG Cluster Lifecycle and SIG Security worked together over several months, inspired from https://github.com/cncf/tag-security self-assessment process to write this self-assessment that identifies feature and documentation that needs work to improve the security posture of the sub-project.

### Note

This is the first of it's kind effort at CNCF where a sub-project of a graduated project was assessed via collaboration between Security experts and maintainers of this sub-project. Past self-assessments via CNCF TAG Security have been focussed on _graduating_ projects. 

If you would like to request self-assessment for your sub-project, please create an issue by clicking on this [link](https://github.com/kubernetes/sig-security/issues/new?assignees=&labels=area%2Fsecurity-assessment%2C+sig%2Fsecurity&template=security-assessment.md&title=REQUEST%3A+Security+Assessment+for+%3Csub-project-name%3E)

Fixes #8 

xref https://github.com/cncf/tag-security/issues/603 and https://github.com/kubernetes-sigs/cluster-api/issues/4446

### Project Tracker

All the issues that are an outcomes of the self-assessment are tracked here: https://github.com/orgs/kubernetes/projects/83/views/1

/sig security cluster-lifecycle
/area security
/kind feature
/cc @sbueringer @Ankitasw @rficcaglia @randomvariable 
/hold for https://github.com/kubernetes/community/pull/6512

sbueringer commented:
I'll try to find some time to go over the doc.

(/cc @fabriziopandini fyi)

neolit123 commented:
> Biggest help needed when reviewing is to look at "status" for each threats

i already provided feedback on some of these while there were in google docs form, i believe.
but more than interested to see what the CAPI maintainers think.

fabriziopandini commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+
+##### Cluster API Core Controller
+
+Cluster API Core controller is a pluggable provider in Cluster API which
```
nit; core controllers are not pluggable

PushkarJ replied:
Updated and shortened to reflect this comment

fabriziopandini commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+
+#### Identified Gaps
+
+A process should be implemented to generate an SBOM for the primary artifacts of
+the Cluster API subprojects, including OCI images and command line binaries.
```
Qs: (sorry if some of them are silly, but 'm learning a lot here)
is there a specific format for SBOM?
what is the expected usage of an SBOM? 
Is there tooling to automatically produce a SBOM?
Do we have an issue in CAPI to track this GAP? if yes, let's link

sbueringer replied:
I think we'll both learn a lot by following folks on that issue: https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/3325 :)

PushkarJ replied:
Since this is covered in STRIDE-TAMPER-1, it makes sense to remove this section in favor of brevity!

sbueringer replied:
Thx for the information. Sounds good tome

fabriziopandini commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+
+#### Identified gaps in development workflow
+
+##### [DEV-01: Outdated OWNERS file](https://github.com/kubernetes-sigs/cluster-api/issues/6151)
```
I'm aware of work in main Kubernetes to ensure the proper hygiene of owner files. As far as you know, Is it possible to leverage on the same tooling?

PushkarJ replied:
Resolving this thread, as the discussions have begun at K8s project level for now and will continue them there

fabriziopandini commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+the [Core Infrastructure Initiative (CII) badging process](https://bestpractices.coreinfrastructure.org/en/projects/5288)
+.
+
+##### DEV-02: Vulnerability report process is poorly defined
```
This is a huge demand for a sub-project without security experts full staffed.
It would be great if there is a template process we can refer to; pushing this a little bit further, as a community does not make sense to join forces (e.g. have a common process, share the same tooling, etc) instead of re-inventing things?

sbueringer replied:
Would be definitely nice to get to a point where an individual project (core CAPI or infra providers) only have to provide a list of contacts.

neolit123 replied:
the idea for private mailing lists for security reports was rejected some time ago, but that was a time when groups were not yet under k8s.io (community managed vs google managed). @PushkarJ i think it might be worth to open a discussion with #sig-k8s-infra about that.

- one private security mailing list per SIG makes sense to me as a start- `sig-foo-security@k8s.io`
- if a subproject really needs its own security it can be `sig-foo-subproject-security@k8s.io`.
- i don't think such a mailing list will be high volume.
- SIG leads + others can be present there and DM maintainers when a security issue arrives.

there is an issue somewhere to change the SECURITY_CONTACTS mechanism.
including github IDs there is pointless as github has no real DMs(?).
there should be a mailing list or personal emails linked either from the file or from the project README.

neolit123 replied:
IIRC, there was a CI job that validates the contents for the SC file.
but you can try adding emails today instead of github IDs as a quick fix.

PushkarJ replied:
100% agree with the comments here and need for repeatable process. The relevant issues on this for k/k have not made much progress afaik (See https://github.com/kubernetes/community/pull/6484)

Maybe for now, it would make sense, to do following:
- Create such a private group for cluster-lifecycle sig as @neolit123 suggested. 
     - PR process to create a group should get sig-k8s-infra's attention if they have any concerns.
- Ensure this mailing list, is listed in the README.md of all the sub-projects owned by this SIG.
     - We ask anyone who is reading to default to `security@kubernetes.io` and 
     - State clearly only emails from SRC members will be addressed / triaged on the private mailing list. 
     - This will ensure we do not have to triage something that has not gone through SRC. 

I will try to take up the org level need for such a mailing list as part of my other responsibilities but out of scope of this assessment. 

Would that work for you all?

PushkarJ replied:
I discussed this in SIG Security meeting today (04/21). SRC is going to have a similar discussion on this topic soon. I think considering the scope of the assessment, having an issue (which we already do) that tracks this work, is good enough for now. Ideally, if the proposed approach goes well it will benefit Cluster API project as well as all other K8s sub-projects and we would feel good about influencing something that had wide impact :)

sbueringer replied:
Sounds good to me!

fabriziopandini commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+
+Status: [Planned](https://github.com/kubernetes-sigs/cluster-api/issues/5398)
+
+##### DEV-03: Project should configure some fuzzing
```
AFAIK done, but I have asked for updates on the issue

sbueringer replied:
Now closed: https://github.com/kubernetes-sigs/cluster-api/issues/6059

fabriziopandini commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+Manager. Access to the management cluster basically provides this admin access
+to all the managed workload clusters and its workloads as well.
+
+#### STRIDE-SPOOF-1 Recommended mitigations
```
I'm not really sure I understand the action required here.
Do we have corresponding issues? should they be created on Providers subprojects vs being in CAPI?

PushkarJ replied:
I think the idea is that we need to add the recommendations that are not implemented in the list below into end user guidance page we will write in https://github.com/kubernetes-sigs/cluster-api/issues/6152 . Does that help resolve the confusion?

fabriziopandini commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+of a cloud service could result in information disclosure and tampering of in
+flight requests relevant to cluster and machine management.
+
+#### STRIDE-SPOOF-2 Recommended Mitigations
```
Same here (I'm not really sure I understand the action required here 😅 )

PushkarJ replied:
No Action needed. This mostly informational

fabriziopandini commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+accounts of maintainers that have privileges and ownership over the code and
+machine image built for Nodes in a cluster.
+
+#### STRIDE-SPOOF-3 Recommended Mitigations
```
As a CAPI project we cannot enforce how providers should manage their release process, but I like the idea of having security guidelines for providers in the CAPI book including this recommendation and others as well.
Would you be so kind to open an issue in CAPI with a list of recommendations we should surface?

PushkarJ replied:
On second thoughts, we may need a separate issue that is tied to Providers only! Will work on opening this one soon.

PushkarJ replied:
Added and linked

fabriziopandini commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+  build action
+  *
+  Status: [Issue filed](https://github.com/kubernetes-sigs/image-builder/issues/803)
+* During install time: Perform mutual authentication when invoking clusterctl
+  command. Verify checksum of the downloaded machine image, before marking them
+  as available. Download with TLS enabled to prevent in-transit tampering.
+  * Status: Partially implemented
+    * Maintainers note: clusterctl by default uses Github APIs for releases
```
Clusterctl doesn't download machine images. this is done by provider's controllers

PushkarJ replied:
Updated to reflect this comment

fabriziopandini commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+#### STRIDE-INFODISCLOSE-2 Recommended Mitigations
+
+* kubeadm token is short lived or deleted
+  * Status: implemented / planned / to be implemented / End User Guidance
```
This is already implemented

fabriziopandini commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+
+* kubeadm token is short lived or deleted
+  * Status: implemented / planned / to be implemented / End User Guidance
+* Disable or restrict SSH access to nodes in a cluster
+  * Status: implemented / planned / to be implemented / End User Guidance
+* Audit KubeConfig files access, edits that are located on the node
+  * Status: implemented / planned / to be implemented / End User Guidance
+* Implement cluster level Pod Security to prevent host mount access to pod
+  * Status: implemented / planned / to be implemented / End User Guidance
```
Q: what is the expectation here (and below)?
Discussing all these points and getting to an agreement in the community/involving all the providers could be challenging, what is the suggested approach to tackle this incrementally?

fabriziopandini reviewed (commented):
@PushkarJ this is great work! thank you!
If I can give a small suggestion is to improve a little bit the report/accompany it with a google sheet or something where it will be easier to have a quick glance at the list of things to do.
Expanding this a little bit, ideally required actions should be prioritized, so we can start planning how to improve CAPI security posture incrementally.

sbueringer commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+
+The general expectation from the control plane controller is to make sure that
+the control plane node is created with the services like etcd, kubernetes API
+server, kubernetes controller manager, scheduler, cloud controller manager,
```
nit: cloud controller manager is not managed by KCP (so at least KCP doesn't fulfill this expectation)

PushkarJ replied:
done

sbueringer commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+
+##### [DEV-01: Outdated OWNERS file](https://github.com/kubernetes-sigs/cluster-api/issues/6151)
+
+Listed maintainers of the Cluster API project are out of date and may continue
```
I think we have to differentiate. The OWNERS file in the CAPI repo only allow /approve + /lgtm

Further access (write/admin access + GCS) is configured in other repositories and they are not all OWNER files (maybe it's just a nitpick :))

For reference. The PRs linked below: https://github.com/kubernetes-sigs/cluster-api/pull/6109#issuecomment-1042567056 should provide a good overview.

PushkarJ replied:
Yeah. Good catch. Updated the section heading to reflect this nuance

sbueringer commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+
+### Spoofing
+
+#### STRIDE-SPOOF-1 - Spoofing of Cloud Admin
```
@mdbooth just fyi. I think that's the point we discussed earlier today in https://kubernetes.slack.com/archives/C8TSNPY4T/p1649253376093879

sbueringer commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+#### STRIDE-SPOOF-4 Recommended Mitigations
+
+* Replace kubeadm node join mechanism with a host attestation mechanism.
+  * Status: Planned
```
Is there something that we can link? (same below)

sbueringer replied:
Linking the CAEP definitely makes sense. I found this issue: https://github.com/kubernetes-sigs/cluster-api/issues/3762 which I think corresponds to the proposal, but they are not linked via GitHub for some reason.

sbueringer commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+  performed using cloud credentials granted to the management and workload
+  cluster’s control plane. Additionally, implement centralized audit collection
+  and alerting for any suspicious activity.
+  * Status: [End user guidance to be written](#cluster-api-end-user-guide)
```
I wonder if we are able to deliver that guidance or rather in which details we can provide it.

Some of the points seem to have a very large scope (I would say some ~ the same size as the whole CAPI project, e.g. it's non trivial to build/deploy/configure/integrate a full-fledged monitoring and auditing solution).

sbueringer commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+
+* kubeadm token is short lived or deleted
+  * Status: implemented / planned / to be implemented / End User Guidance
+* Disable or restrict SSH access to nodes in a cluster
```
Not sure what we can do here. Currently it's basically configurable by the user via KubeadmConfig + some infra providers have additional mechanisms.

sbueringer commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+#### STRIDE-INFODISCLOSE-3 Recommended Mitigations
+
+* Keep CA key/cert pairs short lived or deleted after use from Secrets Manager
+* Shorten CA cert lifetime
```
just to mention it this requires a solution for CA rotation. Otherwise we have a big problem.

sbueringer replied:
Yes

sbueringer commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+#### STRIDE-DOS-2 Recommended Mitigations
+
+* Create safe defaults (soft and hard) for the maximum number of clusters that
+  can be created using a management cluster
```
If the number of clusters is limited they can just scale up a small number of clusters to a very large machine/node count

sbueringer commented on sig-security-assessments/cluster-api/self-assessment.md:
```diff
@@ -0,0 +1,1099 @@
+  creating a cluster by an Infra admin
+  * Status: implemented / planned / to be implemented / End User Guidance
+
+#### STRIDE-DOS-3 - Misuse of Infra Admin credentials to create unlimited number of clusters
```
The title seems to be the same as STRIDE-DOS-2

PushkarJ reviewed (commented):
Thank you again for the exhaustive feedback Stefan, Fabrizio, Lubomir !! Tried to address most comments. Will update the doc soon with relevant changes where needed

PushkarJ commented:
Hi @sbueringer @fabriziopandini 👋🏼 

Just did another walkthrough on the comments and resolved a few with updates. 

Next step: Out of total 58 recommendations, we need to still update status of about 31 of them. Best way to find those 31 would be to:

1. Go here: https://github.com/kubernetes/sig-security/pull/40/files#diff-67678031cfcd98b988ace5dcdbed2b46cec1ed1c68ca2bf48a9b6906662e76ab 
2. Click "Load diff" for `self-assessment.md` file 
3. Search for "Status: implemented / planned / to be implemented / End User Guidance" without double quotes

Once you find them please update select one of the options from:
1. implemented: Please link to a _merged_ PR that implements this
2. planned: Please link to an _open_ issue that implements this
3. to be implemented: Please link to an _unmerged_ PR that implements this
4. End User guidance: Please remove other options in front of status and replace them with `[End user guidance to be written](#cluster-api-end-user-guide)`

Let me know if doing this together in a working session would be helpful and I would be happy to set it up!!

PushkarJ commented:
We are now ready for review. Let me know if I missed linking a tracking issue to any of the threats in the doc.

PushkarJ commented:
/assign aladewberry 

(For review / feedback and overall structure as new lead of security self-assessments)

aladewberry commented:
@PushkarJ Ack! WIll try and get this read through before I'm off next week.

aladewberry commented:
/lgtm 

I like the layout with the description of the project, personas, main flows, and *then* the security examination. I imagine we'll iterate incrementally as we continue assessing other projects, but this seems like an excellent place to start.

PushkarJ commented:
/label tide/merge-method-squash
/remove-hold

aladewberry commented:
/approve

aladewberry commented:
/lgtm

aladewberry commented:
/approve
